go/payments/pkg/cellwise/shard_manager.go

```diff
@@ -66,53 +128,85 @@ func (sm *shardManager) getShards() []AttributionShard {
 
 // add an attributed row to the current shard being built
 func (sm *shardManager) addRowAtt(ctx context.Context, key types.Value, ra rowAtt, raVal types.Value) error {
-	// check and shard if beyond the configured number of rows per shard
-	if sm.rowsInCurrShard >= sm.shardParams.RowsPerShard {
-		err := sm.closeCurrentShard(ctx, key)
+	// if necessary encode the row attribution data as a noms value
+	if raVal == nil {
+		var err error
+		raVal, err = ra.AsValue(sm.nbf, sm.rowAttBuff)
 		if err != nil {
 			return err
 		}
 	}
 
-	// open a new shard if not actively writing a shard
-	if sm.currStore == nil {
-		err := sm.openNewShard(ctx, key)
+	sm.aQ.Push(key, raVal, ra)
+	nextKey := sm.aQ.PeekKey()
+
+	// check and shard if beyond the configured number of rows per shard
+	if sm.rowsBufferred == sm.shardParams.RowsPerShard {
```

Comment:
Would it be safer to do `sm.rowsBufferred >= sm.shardParams.RowsPerShard` ?

Reply:
Yes. But in practice it's the same.  This is checked every time a row is added, so it shouldn't matter... but I made the change.